integration/truncindex_test.go

```diff
@@ -0,0 +1,133 @@
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
+)
+
+func genTruncIndex(normalName string) string {
+	return normalName[:3]
+}
+
+func TestTruncIndex(t *testing.T) {
+	t.Logf("Pull an image")
+	const appImage = "busybox"
+	img, err := imageService.PullImage(&runtimeapi.ImageSpec{Image: appImage}, nil)
+	require.NoError(t, err)
+	imgTruncId := img[:len(img)/2]
+	defer func() {
+		assert.NoError(t, imageService.RemoveImage(&runtimeapi.ImageSpec{Image: imgTruncId}))
+	}()
+
+	t.Logf("Get image status by truncindex, truncId: %s", imgTruncId)
+	res, err := imageService.ImageStatus(&runtimeapi.ImageSpec{Image: imgTruncId})
+	require.NoError(t, err)
+	require.Equal(t, true, (res != nil) && (res.Id != ""))
+	t.Logf("Image status: %v", res)
+
+	// TODO(yanxuean): for failure test case where there are two images with the same truncindex.
+	// if you add n images at least two will share the same leading digit.
+	// "sha256:n" where n is the a number from 0-9 where two images have the same trunc,
+	// for example sha256:9
+	// https://github.com/kubernetes-incubator/cri-containerd/pull/352
+	// I am thinking how I get the two image which have same trunc.
+
+	// TODO(yanxuean): add test case for ListImages
+
+	t.Logf("Create a sandbox")
+	sbConfig := PodSandboxConfig("sandbox", "TruncIndex")
+	sb, err := runtimeService.RunPodSandbox(sbConfig)
+	require.NoError(t, err)
+	sbTruncIndex := genTruncIndex(sb)
+	defer func() {
+		assert.NoError(t, runtimeService.StopPodSandbox(sbTruncIndex))
+		assert.NoError(t, runtimeService.RemovePodSandbox(sbTruncIndex))
+	}()
+
+	t.Logf("Get sandbox status by truncindex")
+	_, err = runtimeService.PodSandboxStatus(sbTruncIndex)
+	require.NoError(t, err)
+
+	t.Logf("Forward port for sandbox by truncindex")
+	_, err = runtimeService.PortForward(&runtimeapi.PortForwardRequest{sbTruncIndex, []int32{80}})
+	require.NoError(t, err)
+
+	// TODO(yanxuean): add test case for ListPodSandbox
+
+	t.Logf("Create a container")
+	cnConfig := ContainerConfig(
+		"containerTruncIndex",
+		appImage,
+		WithCommand("top"),
+	)
+	cn, err := runtimeService.CreateContainer(sbTruncIndex, cnConfig, sbConfig)
+	require.NoError(t, err)
+	cnTruncIndex := genTruncIndex(cn)
+	defer func() {
+		assert.NoError(t, runtimeService.RemoveContainer(cnTruncIndex))
+	}()
+
+	t.Logf("Get container status by truncindex")
+	_, err = runtimeService.ContainerStatus(cnTruncIndex)
+	require.NoError(t, err)
+
+	t.Logf("Start the container")
+	require.NoError(t, runtimeService.StartContainer(cnTruncIndex))
+	defer func() {
+		assert.NoError(t, runtimeService.StopContainer(cnTruncIndex, 10))
+	}()
+
+	t.Logf("Stats the container")
+	_, err = runtimeService.ContainerStats(cnTruncIndex)
+	require.NoError(t, err)
+
+	t.Logf("Update container memory limit after started")
+	err = runtimeService.UpdateContainerResources(cnTruncIndex, &runtimeapi.LinuxContainerResources{
+		MemoryLimitInBytes: 50 * 1024 * 1024,
+	})
+	require.NoError(t, err)
```

Comment:
assert.

Reply:
Are there rule for using assert and require?

Reply:
done

Reply:
Sorry. The rule is that, if the assertion doesn't affect following test, we use `assert`.
If we could not move on without the assertion, we should use `require`.

In your case:
* Here, because no following assertion relying on whether `err == nil` or not, you could use assert. So that we could get as much test result as we can.
* However, if you want to use the return value, e.g. compare the id. You may want to use `require.NoError(t, err)`. Because if not, you'll get a panic accessing `nil` return value, and the whole test suite will fail right away.

So in your case, some of assertion should use `assert`, some should use `require`. You may want to update based on the convention above. The goal is to get as much test result as we can in a run. :)

Reply:
I see. Thank you.